tempesta_fw/http2/buffers.c

```diff
@@ -0,0 +1,1332 @@
+/**
+ *		Tempesta FW
+ *
+ * HTTP/2 bufferization layer for fragment-based parser.
+ *
+ * Copyright (C) 2017 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include <stdint.h>
+#include <string.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include "common.h"
+#include "../pool.h"
+#include "../str.h"
+#include "hash.h"
+#include "errors.h"
+#include "buffers.h"
+
+#define Debug_Buffers 0
+
+#if Debug_Buffers
+#define DPRINTF(...) printf("Buffers: " __VA_ARGS__)
+#define DPUTS(...) puts("Buffers: " __VA_ARGS__)
+#else
+#define DPRINTF(...)
+#define DPUTS(...)
+#endif
+
+/* -------------------------------------------------- */
+/* Input buffers, which are used to handle fragmented */
+/* stings as input of the parsers:		      */
+/* -------------------------------------------------- */
+
+/* Initialize input buffer from the TfwStr: */
+
+void
+buffer_from_tfwstr(HTTP2Input * __restrict p, const TfwStr * __restrict str)
+{
+	p->str = str;
+	p->current = 0;
+	p->n = str->len;
+/* p->tail is initialized here only to avoid */
+/* partial writing into cache line, really */
+/* it does not used by buffer_get() call: */
+	p->tail = 0;
+	p->offset = 0;
+}
```

Comment:
One more string descriptor (`HTTP2Input`) over the already existing one (`TfwStr`) seems too complicated; maybe it is worth to extend TfwStr and apply this logic for it or make separate descriptor.

Reply:
The interfaces are really complicated (seems in aiming to reach generic API for the library). Frankly, I didn't pay enough attention in review of this part of code because it's clear there shouldn't be such data structures.